File: interview_quesitons/stacks_and_queues/4.c
/*
    Queue via Stacks: Implement a MyQueue class which implements a queue using two stacks.
*/

#include "stdio.h"
#include "stdlib.h"
#include "stack.h"


typedef struct {
    struct Stack *oldest;
    struct Stack *newest;

} MyQueue;

MyQueue *MyQueueBuild() {
    MyQueue *queue = malloc(sizeof(MyQueue *));
    queue->oldest = NULL;
    queue->newest = NULL;
    return queue;
}

void myQueueAdd(MyQueue *queue, int val) {
    push(&queue->newest, val);
}

void myQueueShiftStacks(MyQueue *queue) {
    if (isEmpty(queue->oldest)) {
        while (!isEmpty(queue->newest)) {
            push(&queue->oldest, pop(&queue->newest));
        }
    }
}

int myQueuePeek(MyQueue *queue) {
    myQueueShiftStacks(queue);
    return peek(queue->oldest);
}

int myQueueRemove(MyQueue *queue) {
    myQueueShiftStacks(queue);
    return pop(&queue->oldest);
}

int MyQueuePop(MyQueue *queue) {
    return pop(&queue->newest);
}

int MyQueueSize(MyQueue *queue) {
    return getLength(queue->oldest) + getLength(queue->newest);
}

_Bool isMyQueueEmpty(MyQueue *queue) {
    return queue->newest == NULL && queue->oldest == NULL;
}
int main() {
    MyQueue *queue = MyQueueBuild();
    myQueueAdd(queue, 10);
    myQueueAdd(queue, 20);
    myQueueAdd(queue, 30);

    MyQueuePop(queue);
    MyQueuePop(queue);

    printf("%d", isMyQueueEmpty(queue));

}

File: interview_quesitons/stacks_and_queues/4.py
class Queue:
    def __init__(self):
        self.s1 = []
        self.s2 = []

    def add(self, x):

        while self.s1:
            self.s2.append(self.s1.pop())

        self.s1.append(x)

        while self.s2:
            self.s1.append(self.s2.pop())

    def pop(self):

        if not self.s1:
            raise Exception("Q is Empty")

        return self.s1.pop()


q = Queue()
q.add(1)
q.add(2)
q.add(3)

print(q.pop())
print(q.pop())
print(q.pop())

